### PR TITLE
Add search autocomplete API

### LINKED ISSUE #3545: Implement search autocomplete API endpoint
I'm looking to implement the autocomplete for https://github.com/mozilla/addons-frontend/issues/1079 using the `/api/v3/addons/search/` API. I think this would be the right approach over implementing a custom autocomplete API but I could be wrong.

Right now the search API returns a lot of extra info for my needs (I just need name & slug right now, probably icon in the future). Maybe a "condensed" or "_autocomplete" param that reduced the response size (and presumed non-cached request time) would be useful? Not sure about that.

Right now the autocomplete on AMO seems to be separate from the actual search, as the results for the same string I get in the autocomplete are different to the search results page. I'd like to use the search API to avoid this in the future and minimise code paths, but if this is too naive for the kind of responsiveness we'd want for autocomplete let me know.

Opening this for discussion, basically.

### PR DESCRIPTION
**Pull request by diox**
It's based on the regular search API, but fetches less data and pagination is not supported, the number of results will always be 10 max.

Fix #3545

### CODE REVIEW

File: docs/topics/api/addons.rst [hunk 1]
```diff
@@ -79,6 +79,36 @@ This endpoint allows you to search through public add-ons.
...
+.. http:get:: /api/v3/addons/autocomplete/
+
+    :query string q: The search query.
+    :query string app: Filter by :ref:`add-on application <addon-detail-application>` availability.
+    :query string appversion: Filter by application version compatibility. Pass the full version as a string, e.g. ``46.0``. Only valid when the ``app`` parameter is also present.
+    :query string author: Filter by exact author username.
+    :query string category: Filter by :ref:`category slug <category-list>`. ``app`` and ``type`` parameters need to be set, otherwise this parameter is ignored.
+    :query string lang: Activate translations in the specific language for that query. (See :ref:`translated fields <api-overview-translations>`)
+    :query string platform: Filter by :ref:`add-on platform <addon-detail-platform>` availability.
+    :query string tag: Filter by exact tag name. Multiple tag names can be specified, separated by comma(s).
```
Comment (EnTeQuAk) [inline docs/topics/api/addons.rst]: I don't think you need that `(s)` bit, "Multiple tag names" implies that imho.

File: src/olympia/addons/serializers.py [hunk 2]
```diff
@@ -383,8 +383,9 @@ def fake_object(self, data):
         )
 
         # Attach attributes that do not have the same name/format in ES.
-        obj.tag_list = data['tags']
-        obj.disabled_by_user = data['is_disabled']  # Not accurate, but enough.
+        obj.tag_list = data.get('tags', [])
+        obj.disabled_by_user = data.get(
+            'is_disabled', False)  # Not entirely accurate, but enough.
```
Comment (EnTeQuAk) [inline src/olympia/addons/serializers.py]: Move the comment to the line above it and it starts to look way better again imho (no need to break the line)

File: src/olympia/addons/serializers.py [hunk 3]
```diff
@@ -459,6 +460,19 @@ class ESAddonSerializerWithUnlistedData(
     previews = ESPreviewSerializer(many=True, source='all_previews')
 
 
+class ESAddonAutoCompleteSerializer(ESAddonSerializer):
+    class Meta(ESAddonSerializer.Meta):
+        fields = ('id', 'icon_url', 'name', 'url', )
```
Comment (EnTeQuAk) [inline src/olympia/addons/serializers.py]: Not a fan of these trailing commas, unless you're breaking the line as in

```python
fields = (
    'id',
    'icon_url',
    'name',
    'url',
)

File: src/olympia/addons/serializers.py [hunk 4]
```diff
@@ -459,6 +460,19 @@ class ESAddonSerializerWithUnlistedData(
...
+        model = Addon
+
+    def get_url(self, obj):
+        # Addon.get_url_path() wants current_version to exist, but that's just
+        # a safeguard. We don't care and don't want to fetch the current
+        # version field to improve perf, so give it a fake one.
+        obj._current_version = Version()
```
Comment (EnTeQuAk) [inline src/olympia/addons/serializers.py]: just so that I said it: I feel like that should be workaroundable in `current_version` but I don't see an easy way to do that just yet, so 👍 for this here for now. It's documented at least :)

File: src/olympia/addons/views.py [hunk 5]
```diff
@@ -773,6 +774,39 @@ def as_view(cls, **kwargs):
...
+        # excludes a specific set of fields - because we know that autocomplete
+        # only needs to return very few things.
+        included_fields = (
+            'icon_type',  # Needed for icon_url.
+            'id',  # Needed for... id.
+            'modified',  # Needed for icon_url.
+            'name_translations',  # Needed for... name.
+            'persona',  # Needed for icon_url (sadly).
+            'slug',  # Needed for url.
+            'type',  # Needed to attach the Persona for icon_url (sadly).
```
Comment (EnTeQuAk) [inline src/olympia/addons/views.py]: ah, now I understand the persona tests earlier. Maybe they need that same comment or so, I wondered if they belong to a different PR

**Comment by diox** [comment]
> I wonder if we need safeguards to avoid autocomplete queries for, say, < 3 characters or so (I'm guessing that's a client problem though).

That's a good idea, I can look into returning a `400` for this case.

> I don't think people expect or even want to show it the same results as the regular search.

Yeah... I don't know what we want, so I just re-used the same implementation as search, but we can fine-tune this later.
